### ISSUE TITLE
Step-70 runtime error

### ISSUE THREAD

**Issue by Sfiguz7** [dealii/dealii#14844]
While trying to run the step-70 tutorial, the code aborts at runtime due to an assertion:

<details>
<summary>Assertion failure</summary>

> An error occurred in line <902> of file </home/colombo/dealii-candi/tmp/unpack/deal.II-master/include/deal.II/fe/mapping_q_internal.h> in function
    dealii::internal::MappingQImplementation::InverseQuadraticApproximation<dim, spacedim>::InverseQuadraticApproximation(const std::vector<dealii::Point<spacedim> >&, const std::vector<dealii::Point<dim> >&) [with int dim = 2; int spacedim = 2]
The violated condition was:
    ::dealii::deal_II_exceptions::internals::compare_for_equality(real_support_points.size(), unit_support_points.size())
Additional information:
    Dimension 4 not equal to 0.
</details>

A similar issue was encountered in the past in #12029 and #13734, although I am unsure whether this time it is caused by a user error as the dimension of the unit_support_points vector is 0 rather than just a simple mismatch, which would suggest it has not been initialised correctly: that being said, I am positive I have not modified the step-70.cc file and, by backtracking through the stacktrace, I could not find a possible reason for this failure.
I report below my attempt at reconstructing the issue (apologies in advance for the two types of code references, commit specific and master branch). For completeness, I also attach the stacktrace:

<details>
<summary>Stacktrace</summary>
 #0  /home/colombo/dealii-candi/deal.II-master/lib/libdeal_II.g.so.9.5.0-pre: dealii::internal::MappingQImplementation::InverseQuadraticApproximation<2, 2>::InverseQuadraticApproximation(std::vector<dealii::Point<2, double>, std::allocator<dealii::Point<2, double> > > const&, std::vector<dealii::Point<2, double>, std::allocator<dealii::Point<2, double> > > const&)
<span>#</span>1  /home/colombo/dealii-candi/deal.II-master/lib/libdeal_II.g.so.9.5.0-pre: dealii::MappingQ<2, 2>::transform_points_real_to_unit_cell(dealii::TriaIterator<dealii::CellAccessor<2, 2> > const&, dealii::ArrayView<dealii::Point<2, double> const, dealii::MemorySpace::Host> const&, dealii::ArrayView<dealii::Point<2, double>, dealii::MemorySpace::Host> const&) const
<span>#</span>2  /home/colombo/dealii-candi/deal.II-master/lib/libdeal_II.g.so.9.5.0-pre: dealii::Particles::ParticleHandler<2, 2>::sort_particles_into_subdomains_and_cells()
<span>#</span>3  ./step-70: std::enable_if<std::is_convertible<dealii::TrilinosWrappers::MPI::Vector*, dealii::Function<2, double>*>::value==(false), void>::type dealii::Particles::ParticleHandler<2, 2>::set_particle_positions<dealii::TrilinosWrappers::MPI::Vector>(dealii::TrilinosWrappers::MPI::Vector const&, bool)
<span>#</span>4  ./step-70: Step70::StokesImmersedProblem<2, 2>::run()
<span>#</span>5  ./step-70: main

</details>

Here is the extent of my search.
The assertion failure happens here
https://github.com/dealii/dealii/blob/ebe26c794aa824c81297afecfd9ece362922bc33/include/deal.II/fe/mapping_q_internal.h#L902
This is due to the initialisation of an InverseQuadraticApproximation object at
https://github.com/dealii/dealii/blob/master/source/fe/mapping_q.cc#L772
I conclude that here the vector `unit_cell_support`, passed as parameter, is of size 0. This object has been initialised in the MappingQ constructor
https://github.com/dealii/dealii/blob/master/source/fe/mapping_q.cc#L369
In particular, here is the generating function
https://github.com/dealii/dealii/blob/ebe26c794aa824c81297afecfd9ece362922bc33/include/deal.II/fe/mapping_q_internal.h#L218
It is my understanding that the vector would only be empty if renumbering.size() == 0. This happens to reference the parameter `renumber_lexicographic_to_hierarchic` of the MappingQ constructor, which is initialised using
https://github.com/dealii/dealii/blob/d8dacc551ef4a4f8dcfe93361ba5a44ad14b2448/include/deal.II/fe/fe_tools.templates.h#L3182
That is, it inverts the result of this method
https://github.com/dealii/dealii/blob/d8dacc551ef4a4f8dcfe93361ba5a44ad14b2448/include/deal.II/fe/fe_tools.templates.h#L3011
The returned vector is of size `dofs_per_cell = Utilities::fixed_power<dim>(n)`: notice that in our case ` dim == 2, n == degree + 1 == 2` so that dofs_per_cell should indeed be 4: it is here that I stopped as I was not able to figure out what would cause the assertion to fail and find a null-sized vector instead.

Please let me know if you need any additional information, and thank you in advance for the help!

**Comment by bangerth** [comment]
For reference, I see the same error:
```
Running StokesImmersedProblem<2> using Trilinos.
Cycle 0:
Time : 0, time step: 0.002
   Number of degrees of freedom: 9539 (8450+1089 -- 0+0)
Tracer particles: 1313
Solid particles: 9216

--------------------------------------------------------
An error occurred in line <902> of file </home/bangerth/p/deal.II/1/dealii/include/deal.II/fe/mapping_q_internal.h> in function
    dealii::internal::MappingQImplementation::InverseQuadraticApproximation<dim, spacedim>::InverseQuadraticApproximation(const std::vector<dealii::Point<spacedim> >&, const std::vector<dealii::Point<dim> >&) [with int dim = 2; int spacedim = 2]
The violated condition was: 
    ::dealii::deal_II_exceptions::internals::compare_for_equality(real_support_points.size(), unit_support_points.size())
Additional information: 
    Dimension 4 not equal to 0.

Stacktrace:
-----------
#0  /home/bangerth/p/deal.II/1/install/lib/libdeal_II.g.so.9.5.0-pre: dealii::internal::MappingQImplementation::InverseQuadraticApproximation<2, 2>::InverseQuadraticApproximation(std::vector<dealii::Point<2, double>, std::allocator<dealii::Point<2, double> > > const&, std::vector<dealii::Point<2, double>, std::allocator<dealii::Point<2, double> > > const&)
#1  /home/bangerth/p/deal.II/1/install/lib/libdeal_II.g.so.9.5.0-pre: dealii::MappingQ<2, 2>::transform_points_real_to_unit_cell(dealii::TriaIterator<dealii::CellAccessor<2, 2> > const&, dealii::ArrayView<dealii::Point<2, double> const, dealii::MemorySpace::Host> const&, dealii::ArrayView<dealii::Point<2, double>, dealii::MemorySpace::Host> const&) const
#2  /home/bangerth/p/deal.II/1/install/lib/libdeal_II.g.so.9.5.0-pre: dealii::Particles::ParticleHandler<2, 2>::sort_particles_into_subdomains_and_cells()
#3  ./step-70: std::enable_if<std::is_convertible<dealii::TrilinosWrappers::MPI::Vector*, dealii::Function<2, double>*>::value==(false), void>::type dealii::Particles::ParticleHandler<2, 2>::set_particle_positions<dealii::TrilinosWrappers::MPI::Vector>(dealii::TrilinosWrappers::MPI::Vector const&, bool)
#4  ./step-70: Step70::StokesImmersedProblem<2, 2>::run()
#5  ./step-70: main
--------------------------------------------------------
```

Does that trigger anything for anyone who may have touched these functions lately?

**Comment by blaisb** [comment]
Weird... I'll try to take a look at it this week. It's an example I wrote with @luca-heltai

**Comment by peterrum** [comment]
@mschreter Just pointed out that there is an issue that will be fixed by #15007.

**Comment by peterrum** [comment]
@blaisb Did you not convert step-70 to a test. That would resolve https://github.com/dealii/dealii/pull/15007#pullrequestreview-1366774915.
